providers/base/tests/test_gpio_control_test.py

```diff
@@ -0,0 +1,351 @@
+        mock_path.return_value = True
+        mock_path_get.return_value = "1"
+        mock_mapping.return_value = {"1": "32"}
+
+        gpio_controller = GPIOController("1", "1", "out", True)
+        gpio_controller.gpio_node = Path("fake-node")
+        self.assertEqual(gpio_controller.value, "1")
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    def test_set_gpio_value_failed(self, mock_mapping):
+        mock_mapping.return_value = {"1": "32"}
+
+        with self.assertRaises(ValueError):
+            gpio_controller = GPIOController("1", "1", "out", True)
+            gpio_controller.value = "wrong_value"
+
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    @patch("gpio_control_test.GPIOController.off")
+    @patch("gpio_control_test.GPIOController.on")
+    def test_blinking_function(self, mock_on, mock_off, mock_mapping):
+        mock_mapping.return_value = {"1": "32"}
+
+        gpio_controller = GPIOController("1", "1", "out", True)
+        gpio_controller.blinking(1, 0.5)
+        mock_on.assert_called_with()
+        mock_off.assert_called_with()
+
+
+class TestMainFunction(unittest.TestCase):
+
+    @patch("gpio_control_test.GPIOController._write_node")
+    @patch("gpio_control_test.GPIOController._read_node")
+    @patch("gpio_control_test.GPIOController._node_exists")
+    @patch("gpio_control_test.GPIOController.get_gpiochip_mapping")
+    @patch("gpio_control_test.GPIOController.blinking")
+    def test_blinking_test(self, mock_blinking, mock_mapping,
+                           mock_path, mock_read, mock_write):
+        mock_args = Mock(
+            return_value=argparse.Namespace(
+                name="fake-node", duration=5, interval=0.5,
+                gpio_chip="1", gpio_pin="1", need_export=False))
+        mock_mapping.return_value = {"1": "32"}
+        mock_path.return_value = True
+        mock_read.side_effect = ["32", "16", "1", "out"]
+
+        blinking_test(mock_args())
+        mock_blinking.assert_called_once_with(
+            mock_args().duration, mock_args().interval)
+
+    @patch("pathlib.Path.read_text")
+    @patch("pathlib.Path.exists")
+    def test_dump_gpiochip_test(self, mock_path, mock_read):
+        mock_path.return_value = True
+        mock_read.return_value = "mock-string"
+
+        dump_gpiochip(None)
+        mock_path.assert_called_once_with()
+        mock_read.assert_called_once_with()
+
+    @patch("pathlib.Path.exists")
+    def test_dump_gpiochip_test_failed(self, mock_path):
+        mock_path.return_value = False
+
+        with self.assertRaises(FileNotFoundError) as context:
+            dump_gpiochip(None)
+
+        self.assertEqual(
+            str(context.exception), "/sys/kernel/debug/gpio file not exists"
+        )
+
+    def test_led_resource(self):
+        mock_args = Mock(
+            return_value=argparse.Namespace(mapping="DL14:5:1 DL14:5:2"))
+        leds_resource(mock_args())
```

Comment:
```suggestion
        with redirect_stdout(StringIO()) as stdout:
            leds_resource(mock_args())
```

Similar to my other comment, this is to prevent a big chunk of data from being printed out in the unit test logs.